src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -205,59 +184,31 @@ private TagSpanIntervalTree<TTag> GetTagTree(ITextSnapshot snapshot, ImmutableDi
             /// complete almost immediately.  Once open though, our normal delays come into play
             /// so as to not cause a flashy experience.
             /// </summary>
-            private void RecomputeTagsForeground(bool initialTags, bool synchronous)
+            private async Task RecomputeTagsForegroundAsync(bool initialTags, CancellationToken cancellationToken)
             {
                 this.AssertIsForeground();
-                Contract.ThrowIfTrue(synchronous && !initialTags, "synchronous computation of tags is only allowed for the initial computation");
 
                 using (Logger.LogBlock(FunctionId.Tagger_TagSource_RecomputeTags, CancellationToken.None))
                 {
-                    // Stop any existing work we're currently engaged in
-                    _workQueue.CancelCurrentWork();
-
-                    var cancellationToken = GetCancellationToken(initialTags);
+                    // Make a copy of all the data we need while we're on the foreground.  Then switch to a threadpool
+                    // thread to do the computation. Finally, once new tags have been computed, then we update our state
+                    // again on the foreground.
```

Comment:
I was a little confused by this comment until I went down into RecomputeTagsAsync (I couldn't figure out where the switch back to the foreground was).

I wonder if it would make sense to keep all the switches inside this method, then have this call out into different methods that assert their 'threadedness'.  That way it's really easy to see where switching is happening

Reply:
thats' a really good idea.  let me make that in the next PR!